Added authentication with API key and log added for generating credential failure

shah279 opened the pull request:
*Issue #, if available:*

*Description of changes:*
feat: authentication with API key added
feat: unit test cases added for API key auth
fix: Log added for generating credentials failure
refactor: `CryptoUtils.kt` and `AwsAuthorizationHeaderHelper.kt` functions changed to open from internal

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

shah272728 committed:
feat: authentication with API key added
feat: unit test cases added for API key auth
fix: Log added for generating credentials failure
refactor: CryptoUtils.kt and AwsAuthorizationHeaderHelper.kt functions changed to open from internal

ALS-1862

shah272728 committed:
ci: upload-artifact version update

ALS-1862

shah272728 committed:
Merge branch 'refs/heads/main' into ALS-1862

# Conflicts:
#	README.md

olegfilimonov committed:
Merge pull request #21 from makeen-project/ALS-1862

ALS-1862 Authentication with API key added

mbalfour-amzn commented on README.md:
````diff
@@ -35,20 +35,29 @@ You can create an AuthHelper and use it with the AWS Kotlin SDK:
 ```
 // Create a credentail provider using Identity Pool Id with AuthHelper
-private fun exampleCognitoLogin() {
+private suspend fun exampleCognitoLogin() {
````
Nit: "credentail" -> "credential"

shah279 replied:
Fixed

mbalfour-amzn commented on README.md:
```diff
@@ -35,20 +35,29 @@ You can create an AuthHelper and use it with the AWS Kotlin SDK:
 
 // Create a credentail provider using custom credential provider with AuthHelper
-private fun exampleCustomCredentialLogin() {
+private suspend fun exampleCustomCredentialLogin() {
```
"credential"

shah279 replied:
Fixed

mbalfour-amzn commented on README.md:
```diff
@@ -35,20 +35,29 @@ You can create an AuthHelper and use it with the AWS Kotlin SDK:
+
+OR
+
+// Create a credentail provider using Api key with AuthHelper
```
"credential"

shah279 replied:
Fixed

mbalfour-amzn commented on README.md:
```diff
@@ -35,20 +35,29 @@ You can create an AuthHelper and use it with the AWS Kotlin SDK:
+private suspend fun exampleApiKeyLogin() {
+    var authHelper = AuthHelper(applicationContext)
+    var locationCredentialsProvider : LocationCredentialsProvider = authHelper.authenticateWithApiKey("MY-API-KEY", "MY-AWS-REGION")
+    var locationClient = locationCredentialsProvider.getLocationClient()
```
should this be `locationCredentialsProvider?.getLocationClient()`?

shah279 replied:
Fixed

mbalfour-amzn commented on library/src/main/java/software/amazon/location/auth/LocationCredentialsProvider.kt:
```diff
@@ -216,33 +298,29 @@ class LocationCredentialsProvider {
-                    credentials
-                )
+                val credentials = cognitoIdentityClient?.getCredentialsForIdentity(GetCredentialsForIdentityRequest { this.identityId = identityId })
+                    ?.credentials ?: throw Exception("Failed to get  credentials")
```
Nit: "Failed to get credentials" (one space between 'get' and 'credentials')

shah279 replied:
Fixed

mbalfour-amzn commented on library/src/main/java/software/amazon/location/auth/LocationCredentialsProvider.kt:
```diff
@@ -278,29 +357,46 @@ class LocationCredentialsProvider {
         val method = securePreferences.get(METHOD)
-        if (method == "custom" && customCredentials != null) {
+        if ((method == "custom" || method == "apiKey") && customCredentials != null) {
             return customCredentials
```
Should this return empty credentials instead of customCredentials for apiKey?

shah279 replied:
Fixed, code updated

mbalfour-amzn commented on library/src/main/java/software/amazon/location/auth/utils/AwsAuthorizationHeaderHelper.kt:
```diff
@@ -10,7 +10,7 @@ import software.amazon.location.auth.utils.Constants.SIGNING_ALGORITHM
  * Sign the request with the aws authorization header
  */
-internal fun Request.awsAuthorizationHeader(
+fun Request.awsAuthorizationHeader(
```
Why did internal get removed off of these?

shah279 replied:
I thought it would be helpful for our users to access that method but reverted back due to security concerns.

mbalfour-amzn commented on library/src/main/java/software/amazon/location/auth/utils/CryptoUtils.kt:
```diff
@@ -9,19 +9,19 @@ import javax.crypto.Mac
-internal const val ENCODED_CHARACTER_REGEX = "%[0-9A-Fa-f]{2}"
+const val HASHING_ALGORITHM = "SHA-256"
+const val MAC_ALGORITHM = "HmacSHA256"
+const val ENCODED_CHARACTER_REGEX = "%[0-9A-Fa-f]{2}"
```
Why did internal get removed from this file too?

shah272728 committed:
fix: Typo error fixed in README.md
fix: Api key get credential returns empty credential fix
fix: CryptoUtils.kt and AwsAuthorizationHeaderHelper.kt files method revert back to internal

ALS-1862

olegfilimonov committed:
Merge pull request #22 from makeen-project/ALS-1862
